### ISSUE TITLE
Optimize build for easy and fast building on Travis

### ISSUE THREAD

**Issue by probonopd** [AppImage/AppImageKit#529]
The current build artefacts are optimized for compatibility with a wide area of systems, in order to prevent users of AppImageKit from having to build from source. With more and more users using our tools as part of a contiuous build chain on Travis CI, some projects ask to build everything from source.

For example, users like Microsoft and Audacity may not be happy to pull precompiled code from a third party such as us, even though all of that code is open source and built on Travis CI in the open.

We could offer, in addition to what we have now, a second build path that is optimized for easy and fast building on Travis CI. I think this would look greatly different from what we currently have, since we would probably need no Docker, need to do much less building of common third-party parts, and would not have to do as much semi-static linking and testing. The objective of this second build path which should cover both `appimagetool` and `linuxdeployqt` would be to get a functional `appimagetool` and `linuxdeployqt` compiled on Travis CI in the shortest amount of time possible, while having to run nowhere else but on that very instance of Travis CI. I think it would also be fine that the runtime, when built in this way, would not be compatible with older systems, since the content of the AppImage built this way by the AppImageKit user would equally not run on older systems. As an extra measure, we could even put in a check so that `appimagetool` and `linuxdeployqt` built this way would refuse to run on any other system.

**Comment by TheAssassin** [comment]
I'd finish the CMake configuration, and use that for these purposes. It already builds appimagetool in acceptable time (even with the external sources). We could add the remaining executables (appimaged etc.) and check how long builds take with CMake in comparison to the shell script.

**Comment by probonopd** [comment]
Yes. In fact we could probably make a special CMakeFiles.txt for this purpose, which could possibly be much, much simpler.

**Comment by TheAssassin** [comment]
@probonopd we might have to perform some requirements engineering and build the CMake configuration from that list instead of trying to mimic the existing system. A clear begin is the easiest way to implement a proper CMake configuration in this case.

**Comment by Calinou** [comment]
What about telling people to pre-build their own `appimagetool` or `linuxdeployqt`, and download their self-built binary in their CI pipeline? So that the pipeline doesn't spend any time compiling it.

**Comment by probonopd** [comment]
That may be just too much hassle for an organization like Microsoft.

**Comment by TheAssassin** [comment]
Possibly solved by #568.

**Comment by TheAssassin** [comment]
Using pre-built Docker images which bundle all the dependencies, with a custom built compiler toolchain and (some) prebuilt dependencies reduces the build times by 40-50% on average. https://github.com/AppImage/AppImageBuild

**Comment by probonopd** [comment]
While this is a great achievement @TheAssassin the focus of this ticket is a different one: Take a company that considers to use AppImage to distribute their application. They don't want to add a lot of dependencies to the outside, and want to clone (and assess) all sources that are needed to build their product. I have heard this from multiple ones. The least thing they are gonna want is a "fat outside dependency" in the form of a Docker container.

We are using CentOS 6 to build this project in order to make the most compatible binaries we can. Many application projects don't go through the same hassle, and just build for Ubuntu 14.04 or such. For this specific use case, not all the fancy stuff we are doing in this repository is needed. Basically they just need a straightforward way to build `runtime` and `appimagetool` with the minimal outside deps possible.

**Comment by TheAssassin** [comment]
> Basically they just need a straightforward way to build runtime and appimagetool with the minimal outside deps possible.

I don't know how we could solve this problem, unless we add the _entire_ sources in the Git repository, which is far from ideal. We could also think about an option to release "self-contained" source tarballs, containing all the other sources, with a patched build system using those then, rather than the external resources. We could code such an option into the existing build system, and maintain it next to our primary build system. I could imagine a feature like "if tarball/directory called <library name> is available in place xyz, use that instead of downloading the source".

**Comment by probonopd** [comment]
> Edit: turned out to be as simple as passing a path to a tarball rather than the URL to it, so all we'd need to do when shipping such tarballs is to replace the URLs by absolute paths.
I think we can automate that even. Should we generate such tarballs for continuous builds?

Yes, I think this would be useful for some projects, but let's hear from actual users whether there is interest in this.

**Comment by TheAssassin** [comment]
We can also just throw in these tarballs by default, and avoid downloading files _at all_.

**Comment by probonopd** [comment]
Adding tarballs to a git repo isn't "nice". Better add the tarballs to archives when we do releases, but let's hear from actual users whether there is interest in this.

**Comment by TheAssassin** [comment]
@probonopd the idea behind this is, we can't know whether one day these sources will be deleted from the servers.

But I have an idea how to solve that, too. We can probably just create a separate Git repository, similar to AppImageBuild, and mirror them there, so we're on the safe side. Heck, we could even use that as a submodule then.

**Comment by hideout** [comment]
good point

2018-02-03 12:55 GMT+01:00 TheAssassin <notifications@github.com>:
